### PR TITLE
APP-2996: Datafeed ID repository for Datafeed Service V1

### PR DESCRIPTION
**Pull request by symphony-hong**
Create an interface DatafeedRepository for handling storing
a datafeed id which is created by using DatafeedServiceV1.

Create OnDiskDatafeedRepository implementing the DatafeedRepository
for storing datafeed id on disk.

In DatafeedServiceV1, users can use the OnDiskDatafeedRepository
implementation by default or we can inject their own repository, for
example: MongoDB, SQL storage...

Update unittest

### CODE REVIEW

**Review by symphony-elias — APPROVED** [review]
LGTM

**Review by symphony-youri — COMMENTED** [review]
Do we have users that need that? To me it looks like the datafeed id is something quite volatile and that storing it on disk is enough?

File: symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/datafeed/DatafeedRepository.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+   * @param datafeedId the datafeed id to be persisted.
+   */
+  void write(String datafeedId);
+
+  /**
+   * Read the persisted datafeed id from the storage.
+   *
+   * @return The persisted datafeed id.
+   */
+  String read();
```
Comment (symphony-thibault) [inline symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/datafeed/DatafeedRepository.java:21]: Could we make this `read` method return an `Optional`?

File: symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/datafeed/DatafeedRepository.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package com.symphony.bdk.core.service.datafeed;
+
+/**
+ * A repository interface for storing a datafeed id.
+ * By using DatafeedServiceV1, the created datafeed id should be persisted manually on the BDK side.
+ */
+public interface DatafeedRepository {
```
Comment (symphony-thibault) [inline symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/datafeed/DatafeedRepository.java:7]: `DatafeedIdRepository`

File: symphony-bdk-core/src/test/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedRepositoryTest.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+public class OnDiskDatafeedRepositoryTest {
```
Comment (symphony-thibault) [inline symphony-bdk-core/src/test/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedRepositoryTest.java:24]: Add javadoc comment: 
```
/**
 * Test class for the {@link OnDiskDatafeedRepository}.
 */
```

**Comment by symphony-thibault** [comment]
> @symphony-youri Do we have users that need that? 

Apart from me: no :) But I already know that it would be interesting for the Integration team that already works with MongoDB on on-prem side (future `MongoDatafeedIdRepository`?). 

> To me it looks like the datafeed id is something quite volatile and that storing it on disk is enough?

The datafeed id is a critical information. Loosing it can lead to a loss of messages, and potential business impact for our customers. As well, I'm not sure that the disk strategy works well in a Docker environment, to be confirmed.

Last argument: separation of concerns, simplification of the `DatafeedServiceV1` implementation. Code is cleaner an easier to maintain (pure personal POV though).

**Comment by opoupeney** [comment]
The disk strategy works for most of our customers although it requires to mount a persistent volume in your container (not ideal), that's why the Integration team is looking for a DB alternative (which has its own constraints with the highly secure authentication some customers have)

File: symphony-bdk-core/src/test/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepositoryTest.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Optional;
+
+/**
+ * Test class for the {@link OnDiskDatafeedIdRepository}.
```
Comment (symphony-youri) [inline symphony-bdk-core/src/test/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepositoryTest.java:27]: The comment here is not needed IMHO

File: symphony-bdk-core/src/test/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepositoryTest.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  void setUp(@TempDir Path tempDir) throws BdkConfigException {
+    this.bdkConfig = BdkConfigLoader.loadFromClasspath("/config/config.yaml");
+    BdkDatafeedConfig datafeedConfig = this.bdkConfig.getDatafeed();
+    datafeedConfig.setIdFilePath(tempDir.toString());
+    this.bdkConfig.setDatafeed(datafeedConfig);
+    this.datafeedFile = tempDir.resolve("datafeed.id");
+  }
+
+  @Test
+  void writeDatafeedTest() throws IOException {
```
Comment (symphony-youri) [inline symphony-bdk-core/src/test/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepositoryTest.java:44]: tests are easier to read when steps (Given/When/Then) are separated with line breaks IMHO

File: symphony-bdk-core/src/test/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepositoryTest.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  void readEmptyDatafeedFileTest() throws IOException {
+    FileUtils.writeStringToFile(datafeedFile.toFile(), "", StandardCharsets.UTF_8);
+    DatafeedIdRepository repository = new OnDiskDatafeedIdRepository(bdkConfig);
+    Optional<String> datafeedId = repository.read();
+    assertFalse(datafeedId.isPresent());
+  }
+
+  @Test
+  void readNotExistDatafeedFileTest() {
+    DatafeedIdRepository repository = new OnDiskDatafeedIdRepository(bdkConfig);
```
Comment (symphony-youri) [inline symphony-bdk-core/src/test/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepositoryTest.java:79]: I would perhaps move the init of the repo to the setup as we always do it the same way

File: symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepository.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+      List<String> lines = Files.readAllLines(datafeedIdPath);
+      if (lines.isEmpty() || !lines.get(0).contains("@")) {
+        return Optional.empty();
+      }
+      String[] persistedDatafeed = lines.get(0).split("@");
+      datafeedId = persistedDatafeed[0];
+      log.info("Retrieved datafeed id from datafeed repository: {}", datafeedId);
+      return Optional.of(datafeedId);
+    } catch (IOException e) {
+      log.debug("No persisted datafeed id could be retrieved from disk");
```
Comment (symphony-youri) [inline symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepository.java:65]: in that case is it ok to hide the exception or should we propagate it?
if it were a database as the repository would it change?

**Comment by symphony-youri** [comment]
> 
> Last argument: separation of concerns, simplification of the `DatafeedServiceV1` implementation. Code is cleaner an easier to maintain (pure personal POV though).

We could do that having a public interface and adding an extension point in the SDK


Isn't going to be quickly obsolete with DF2? Or is going to take a long time to have it deployed everywhere?

**Comment by symphony-youri** [comment]
> 
> The datafeed id is a critical information. Loosing it can lead to a loss of messages, and potential business impact for our customers. As well, I'm not sure that the disk strategy works well in a Docker environment, to be confirmed.


Does it mean we can only have one DF id for a bot? for me it was more like a session token given that we sometime recreate it
Of for instance what happens if we run two instances of the bot with different ids (or the same id?)

**Comment by opoupeney** [comment]
From an agent point of view, a bot can have up to 5 dafateed opened (to be verified though)

File: symphony-bdk-core/src/test/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepositoryTest.java (hunk 4 above)
Comment (symphony-thibault) [reply symphony-bdk-core/src/test/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepositoryTest.java:27]: I like it (personally), quite convenient when navigating through the code. You can easily jump to the tested class by CTRL+Click on the link. Could be optional, I agree.

File: symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepository.java (hunk 7 above)
Comment (symphony-hong) [reply symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepository.java:65]: I'm not sure if it's needed to be logged or not, so I just log it as debug log. But at least, I think the bot developers should be notified if no datafeed id is retrieved.

File: symphony-bdk-core/src/test/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepositoryTest.java (hunk 4 above)
Comment (symphony-youri) [reply symphony-bdk-core/src/test/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepositoryTest.java:27]: command + shit + t :)

File: symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepository.java (hunk 7 above)
Comment (symphony-youri) [reply symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepository.java:65]: For instance let's say it is a database and the database is not available, could we end creating datafeed indefinitely?
Or that we don't have write permissions on the file?

**Review by symphony-youri — APPROVED** [review]

File: symphony-bdk-core/src/test/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepositoryTest.java (hunk 4 above)
Comment (symphony-thibault) [reply symphony-bdk-core/src/test/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepositoryTest.java:27]: Screenshot 2020-08-28 at 18 56 54

Did not know this one !

**Review by symphony-thibault — COMMENTED** [review]
LGTM 👍

File: symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepository.java (hunk 7 above)
Comment (symphony-hong) [reply symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepository.java:65]: Oh so, you're talking about the one in write method. Maybe we should propagate the exception? But if we throws IOException, then we might have problem when we implement other kinds of repository. Maybe defining a dedicated exception might work?

Comment (symphony-youri) [reply symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/datafeed/impl/OnDiskDatafeedIdRepository.java:65]: I'd say both read/write could raise exceptions, IOException would be fine by me

**Pull request merged**